guidance/llms/_llama_cpp.py

```diff
@@ -0,0 +1,724 @@
+    use_mmap: bool = False
+    verbose: bool = False
+
+class LlamaCpp(LLM):
+    """ A HuggingFace transformers language model with Guidance support.
+    """
+
+    cache = LLM._open_cache("_llama_cpp.diskcache")
+
+    def __init__(self, settings=LlamaCppSettings(), caching=True, token_healing=False, acceleration=False,
+                 temperature=0.25,
+                 role_start=None, role_end=None):
+        super().__init__()
+
+        from llama_cpp import Llama, llama_n_vocab
+        self.settings = settings
+        self.model_obj = Llama(
+            settings.model,
+            n_gpu_layers=settings.n_gpu_layers,
+            f16_kv=settings.f16_kv,
+            use_mlock=settings.use_mlock,
+            embedding=settings.embedding,
+            n_threads=settings.n_threads,
+            n_batch=settings.n_batch,
+            n_ctx=settings.n_ctx,
+            last_n_tokens_size=settings.last_n_tokens_size,
+            logits_all=settings.logits_all,
+            use_mmap=settings.use_mmap,
+            verbose=settings.verbose
+        )
+        self.device = None
+        self.vocab_size = llama_n_vocab(self.model_obj.ctx)
+        self._generate_call = self.model_obj.create_completion
+        base_name = os.path.basename(settings.model)  # get the name of the file
+        file_name_without_extension, _ = os.path.splitext(base_name)
+        self.model_name = file_name_without_extension
+        self.caching = caching
+        self.current_time = time.time()
+        self.call_history = collections.deque()
+        self.temperature = temperature
+        self.token_healing = token_healing
+        self.acceleration = acceleration
+        temp = self.model_obj.tokenize("temp".encode(), True)
+        self._prefix_ids = [temp[0], 100]  # token ids that we use to decode tokens after a prefix
+        self._prefix_str = self.model_obj.detokenize(tokens=self._prefix_ids).decode("utf-8")
+
+        self._token_prefix_map = self._build_token_prefix_map()
+
+    def prefix_matches(self, prefix):
+        """ Return the list of tokens that match the given prefix.
+        """
+        return [v for arr in self._token_prefix_map.values(prefix=prefix) for v in arr]
+
+    def encode(self, string, fragment=True, **kwargs):
+
+        if isinstance(string, bytes):
+            string = string.decode("utf-8")
+
+        if fragment:
+            string = self._prefix_str + string
+        out = {}
+        if "return_tensors" in kwargs:
+            out['input_ids'] = self.model_obj.tokenize(string.encode("utf-8"))
+        else:
+            out['input_ids'] = self.model_obj.tokenize(string.encode('utf-8'))
+
+        # remove the start token when we are encoding a suffix
+        if fragment:
+            if out['input_ids'][1] == self.model_obj.token_bos():  # sometime the tokenizer adds an extra start token
+                out = out['input_ids'][3:]
+            else:
+                out = out['input_ids'][2:]
```

Comment:
Thanks for your fantastic work, @Maximilian-Winter!

While testing this PR + [https://github.com/abetlen/llama-cpp-python/pull/252](https://github.com/abetlen/llama-cpp-python/pull/252) I run into some issues.

Reply:
Could you test the notebook again? It should now work!

Reply:
Thanks! Generating into variables with patterns and stops works well now.

However, guidance `select` doesn't seem to work for me yet. Here's what I'm running:

```python
import guidance

# Create a LlamaCppSettings instance and set the model path and tokenizer name.
settings = guidance.llms.LlamaCppSettings()
settings.model = "../llama.cpp/models/TheBloke/WizardLM-7B-uncensored-GGML/WizardLM-7B-uncensored.ggmlv3.q5_1.bin"
# Hugginface original tokenizer needed at the moment.
settings.tokenizer_name = "ehartford/WizardLM-7B-Uncensored"

# Create a LlamaCpp instance and pass the settings to it.
llama = guidance.llms.LlamaCpp(settings=settings, caching=False)

# we can pre-define valid option sets
valid_weapons = ["sword", "axe", "mace", "spear", "bow", "crossbow"]

# define the prompt
character_maker = guidance("""The following is a character profile for an RPG game in JSON format.
```json
{
    "id": "{{id}}",
    "description": "{{description}}",
    "name": "{{gen 'name' temperature=0.7}}",
    "age": {{gen 'age' pattern='[0-9]+' max_tokens=3 stop=','}},
    "weapon": "{{select 'weapon' options=valid_weapons}}"
}```""")

# generate a character
char = character_maker(
    id="e1f491f7-7ab8-4dac-8c20-c92b5e7d883d",
    description="A quick and nimble fighter.",
    valid_weapons=valid_weapons, llm=llama
)
```
This fills out the `name` and `age` vars nicely but upon running the select exits with the following error:

```bash
...
File ~/.../guidance/guidance/_program_executor.py:429, in ProgramExecutor.visit(self, node, next_node, next_next_node, prev_node, parent_node, grandparent_node)
    427     else:
    428         inner_prev_node = prev_node
--> 429     visited_children.append(await self.visit(child, inner_next_node, inner_next_next_node, inner_prev_node, node, parent_node))
    430 # visited_children = [self.visit(child) for child in node.children]
    432 if len(visited_children) == 1:

File ~/.../guidance/guidance/_program_executor.py:292, in ProgramExecutor.visit(self, node, next_node, next_next_node, prev_node, parent_node, grandparent_node)
    290 if inspect.iscoroutinefunction(command_function):
    291     await asyncio.sleep(0) # give other coroutines a chance to run
--> 292     command_output = await command_function(*positional_args, **named_args)
    293 else:
    294     command_output = command_function(*positional_args, **named_args)

File ~/.../guidance/guidance/library/_select.py:164, in select(variable_name, options, logprobs, list_append, _parser_context)
    161         parser.set_variable(logprobs, option_logprobs)
    163 if max(option_logprobs.values()) <= -1000:
--> 164     raise ValueError("No valid option generated in #select! Please post a GitHub issue since this should not happen :)")
    166 partial_output(selected_option)

ValueError: No valid option generated in #select! Please post a GitHub issue since this should not happen :)
```

Using the debugger I can confirm that indeed all `option_logprobs.values()` are <= -1000:
```python
option_logprobs.values()
dict_values([-1000, -1000, -1000, -1000, -1000, -1000])
```

Is there any other information I can give you to help resovle the issue?

Reply:
Here the output with another model and tokenizer:
The following is a character profile for an RPG game in JSON format.
```json
{
    "id": "e1f491f7-7ab8-4dac-8c20-c92b5e7d883d",
    "description": "A quick and nimble fighter.",
    "name": "Garrett",
    "age": 6,
    "weapon": "sword"
}```

Reply:
Do you also have the latest commit of my guidance fork? (Sorry for asking just want to be sure)

Reply:
Yeah, I'm running the following version:
```bash
❯ git --no-pager log -n 1
commit 737a5f460849e290854378060487d80545888780 (HEAD -> my_llama_cpp, Maximilian-Winter/main)
Author: Maximilian-Winter <maximilian.winter.91@gmail.com>
Date:   Thu May 25 00:49:06 2023 +0200

    Cleaned up torch import.
```

What seems to happen is that at [guidance/library/_select.py#LL120C16-L120C16](https://github.com/microsoft/guidance/blob/7d31c803ead08f9a7ee613f6514ed01a895ee65c/guidance/library/_select.py#LL120C16-L120C16) `top_logprobs` gets truncated to `{21322: -0.4414818676868545}` which is the token id for `'Short'`. Subsequently `sub_logprobs` evaluates to `{}` and therefore all options stay with their default value of `-1000`.

Reply:
Does this happen because guidance doesn't find `current_prefix + 'Shorts'` in the generated `token_map` (because 'Shorts' simply isn't an option) and therefore short-circuits?

Reply:
I ran the code from you and it works for me.

Reply:
Will have to check further.

Reply:
`import guidance
# Create a LlamaCppSettings instance and set the model path and tokenizer name.
settings = guidance.llms.LlamaCppSettings()
settings.model = "../ggml-v3-models/WizardLM-7B-uncensored.ggmlv3.q5_1.bin"
# Hugginface original tokenizer needed at the moment.
settings.tokenizer_name = "ehartford/WizardLM-7B-Uncensored"

# Create a LlamaCpp instance and pass the settings to it.
llama = guidance.llms.LlamaCpp(settings=settings, caching=False)

# we can pre-define valid option sets
valid_weapons = ["sword", "axe", "mace", "spear", "bow", "crossbow"]

# define the prompt
character_maker = guidance("""The following is a character profile for an RPG game in JSON format.
```json
{
    "id": "{{id}}",
    "description": "{{description}}",
    "name": "{{gen 'name' temperature=0.7}}",
    "age": {{gen 'age' pattern='[0-9]+' max_tokens=3 stop=','}},
    "weapon": "{{select 'weapon' options=valid_weapons}}"
}```""")

# generate a character
char = character_maker(
    id="e1f491f7-7ab8-4dac-8c20-c92b5e7d883d",
    description="A quick and nimble fighter.",
    valid_weapons=valid_weapons, llm=llama
)

print(char)`

Reply:
I only had this problem before I assigned logits properly.

Reply:
But I fixed that a couple of hours ago!

Reply:
```` python
import guidance
# Create a LlamaCppSettings instance and set the model path and tokenizer name.
settings = guidance.llms.LlamaCppSettings()
settings.model = "../ggml-v3-models/WizardLM-7B-uncensored.ggmlv3.q5_1.bin"
# Hugginface original tokenizer needed at the moment.
settings.tokenizer_name = "ehartford/WizardLM-7B-Uncensored"

# Create a LlamaCpp instance and pass the settings to it.
llama = guidance.llms.LlamaCpp(settings=settings, caching=False)

# we can pre-define valid option sets
valid_weapons = ["sword", "axe", "mace", "spear", "bow", "crossbow"]

# define the prompt
character_maker = guidance("""The following is a character profile for an RPG game in JSON format.
```json
{
    "id": "{{id}}",
    "description": "{{description}}",
    "name": "{{gen 'name' temperature=0.7}}",
    "age": {{gen 'age' pattern='[0-9]+' max_tokens=3 stop=','}},
    "weapon": "{{select 'weapon' options=valid_weapons}}"
}```""")

# generate a character
char = character_maker(
    id="e1f491f7-7ab8-4dac-8c20-c92b5e7d883d",
    description="A quick and nimble fighter.",
    valid_weapons=valid_weapons, llm=llama
)

print(char)
```

Reply:
> I only had this problem before I assigned logits properly.

Let me make sure I'm really running the up-to-date version of your fork of llama-cpp-python

Reply:
Yeah, my bad! I was on the latest commit, but my installation was probably stale. After uninstalling and reinstalling, the above code works now.

Thank you so much for your patience and great work!